registry/base_registry.go

```diff
@@ -0,0 +1,358 @@
+	CloseAndNilClient()
+	CloseListener()
+	InitListeners()
+}
+
+type BaseRegistry struct {
+	context             context.Context
+	facadeBasedRegistry FacadeBasedRegistry
+	*common.URL
+	birth int64          // time of file birth, seconds since Epoch; 0 if unknown
+	wg    sync.WaitGroup // wg+done for zk restart
+	done  chan struct{}
+
+	cltLock  sync.Mutex
+	services map[string]common.URL // service name + protocol -> service config
+}
+
+func (r *BaseRegistry) InitBaseRegistry(url *common.URL, facadeRegistry FacadeBasedRegistry) Registry {
+	r.URL = url
+	r.birth = time.Now().UnixNano()
+	r.done = make(chan struct{})
+	r.services = make(map[string]common.URL)
+	r.facadeBasedRegistry = facadeRegistry
+	r.wg.Add(1)
+	return r
+}
+
+func (r *BaseRegistry) GetUrl() common.URL {
+	return *r.URL
+}
+func (r *BaseRegistry) Destroy() {
+	r.facadeBasedRegistry.CloseListener()
+	close(r.done)
+	r.wg.Wait()
+	r.closeRegisters()
+}
+
+func (r *BaseRegistry) Register(conf common.URL) error {
+	var (
+		ok  bool
+		err error
+	)
+	role, _ := strconv.Atoi(r.URL.GetParam(constant.ROLE_KEY, ""))
+	switch role {
+	case common.CONSUMER:
+		r.cltLock.Lock()
+		_, ok = r.services[conf.Key()]
+		r.cltLock.Unlock()
+		if ok {
+			return perrors.Errorf("Path{%s} has been registered", conf.Path)
+		}
+
+		err = r.register(conf)
+		if err != nil {
+			return perrors.WithStack(err)
+		}
+
+		r.cltLock.Lock()
+		r.services[conf.Key()] = conf
+		r.cltLock.Unlock()
+		logger.Debugf("(consumerRegistry)Register(conf{%#v})", conf)
+
+	case common.PROVIDER:
+
+		// Check if the service has been registered
+		r.cltLock.Lock()
+		// Note the difference between consumer and consumerZookeeperRegistry (consumer use conf.Path).
+		// Because the consumer wants to provide monitoring functions for the selector,
+		// the provider allows multiple groups or versions of the same service to be registered.
+		_, ok = r.services[conf.Key()]
+		r.cltLock.Unlock()
+		if ok {
+			return perrors.Errorf("Path{%s} has been registered", conf.Key())
+		}
+
+		err = r.register(conf)
+		if err != nil {
+			return perrors.WithMessagef(err, "register(conf:%+v)", conf)
+		}
+
+		r.cltLock.Lock()
+		r.services[conf.Key()] = conf
+		r.cltLock.Unlock()
+
+		logger.Debugf("(ProviderRegistry)Register(conf{%#v})", conf)
+	}
+
+	return nil
+}
+
+func (r *BaseRegistry) service(c common.URL) string {
+	return url.QueryEscape(c.Service())
+}
+
+func (r *BaseRegistry) RestartCallBack() bool {
+
+	// copy r.services
+	services := []common.URL{}
+	for _, confIf := range r.services {
+		services = append(services, confIf)
+	}
+
+	flag := true
+	for _, confIf := range services {
+		err := r.register(confIf)
+		if err != nil {
+			logger.Errorf("(ZkProviderRegistry)register(conf{%#v}) = error{%#v}",
+				confIf, perrors.WithStack(err))
+			flag = false
+			break
+		}
+		logger.Infof("success to re-register service :%v", confIf.Key())
+	}
+	r.facadeBasedRegistry.InitListeners()
```

Comment:
There is no such step in etcd logic, so you need to check it

Reply:
> There is no such step in etcd logic, so you need to check it
Which step?